sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteRename.java

```diff
@@ -95,237 +83,376 @@ public WriteRename(
-
-    BigQueryHelpers.PendingJob retryJob =
-        startCopy(
-            bqServices.getJobService(c.getPipelineOptions().as(BigQueryOptions.class)),
-            getDatasetService(c.getPipelineOptions().as(BigQueryOptions.class)),
-            jobIdPrefix,
-            finalTableDestination.getTableReference(),
-            tempTables,
-            writeDisposition,
-            createDisposition,
-            kmsKey,
-            loadJobProjectId);
-    return new PendingJobData(retryJob, finalTableDestination, tempTables, window);
-  }
+    private PendingJobData startWriteRename(
+        TableDestination finalTableDestination,
+        Iterable<WriteTables.Result> tempTableNames,
+        ProcessContext c,
+        BoundedWindow window)
+        throws Exception {
+      // The pane may have advanced either here due to triggering or due to an upstream trigger. We
+      // check the upstream
+      // trigger to handle the case where an earlier pane triggered the single-partition path. If
+      // this
+      // happened, then the
+      // table will already exist so we want to append to the table.
+      WriteTables.@Nullable Result firstTempTable = Iterables.getFirst(tempTableNames, null);
+      boolean isFirstPane =
+          firstTempTable != null && firstTempTable.isFirstPane() && c.pane().isFirst();
+      WriteDisposition writeDisposition =
+          isFirstPane ? firstPaneWriteDisposition : WriteDisposition.WRITE_APPEND;
+      CreateDisposition createDisposition =
+          isFirstPane ? firstPaneCreateDisposition : CreateDisposition.CREATE_NEVER;
+      List<TableReference> tempTables =
+          StreamSupport.stream(tempTableNames.spliterator(), false)
+              .map(
+                  result ->
+                      BigQueryHelpers.fromJsonString(result.getTableName(), TableReference.class))
+              .collect(Collectors.toList());
 
-  private BigQueryHelpers.PendingJob startCopy(
-      JobService jobService,
-      DatasetService datasetService,
-      String jobIdPrefix,
-      TableReference ref,
-      List<TableReference> tempTables,
-      WriteDisposition writeDisposition,
-      CreateDisposition createDisposition,
-      @Nullable String kmsKey,
-      @Nullable ValueProvider<String> loadJobProjectId) {
-    JobConfigurationTableCopy copyConfig =
-        new JobConfigurationTableCopy()
-            .setSourceTables(tempTables)
-            .setDestinationTable(ref)
-            .setWriteDisposition(writeDisposition.name())
-            .setCreateDisposition(createDisposition.name());
-    if (kmsKey != null) {
-      copyConfig.setDestinationEncryptionConfiguration(
-          new EncryptionConfiguration().setKmsKeyName(kmsKey));
+      List<String> tempTableStrings =
+          StreamSupport.stream(tempTableNames.spliterator(), false)
+              .map(Result::getTableName)
+              .collect(Collectors.toList());
+      ;
```

Comment:
Why do we convert from TableReference to strings here then back to TableReference in `TempTableCleanupFn`? Might be worth adding a comment here to clarify